packages/theia-integration/src/browser/diagram/glsp-diagram-widget.ts

```diff
@@ -201,6 +213,80 @@ export class GLSPDiagramWidget extends DiagramWidget implements SaveableSource {
     protected async clearGlobalSelection(): Promise<void> {
         this.theiaSelectionService.selection = undefined;
     }
+
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    storeState(): object {
+        // the viewport is stored in the application layout
+        // so there is no need to keep it in the storage
+        this.removeViewportDataFromStorageService();
+        return { ...super.storeState(), ...this.getViewportData() };
+    }
+
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    restoreState(oldState: object): void {
+        super.restoreState(oldState);
+        if (isViewportDataContainer(oldState)) {
+            this.setViewportData(oldState);
+        }
+    }
+
+    protected storeViewportDataInStorageService(): void {
+        if (!this.storeViewportStateOnClose) {
+            return;
+        }
+        const viewportData = this.getViewportData();
+        if (viewportData) {
+            this.storage.setData<ViewportDataContainer>(this.viewportStorageId, viewportData);
+        }
+    }
+
+    protected async restoreViewportDataFromStorageService(): Promise<void> {
+        if (!this.storeViewportStateOnClose) {
+            return;
+        }
+        const viewportData = await this.storage.getData<ViewportDataContainer>(this.viewportStorageId);
+        if (viewportData) {
+            this.setViewportData(viewportData);
+        }
+    }
+
+    protected async removeViewportDataFromStorageService(): Promise<void> {
+        return this.storage.setData<ViewportDataContainer | undefined>(this.viewportStorageId, undefined);
+    }
+
+    protected get viewportStorageId(): string {
+        return this.options.diagramType + ':' + this.options.uri;
+    }
+
+    protected getViewportData(): ViewportDataContainer | undefined {
+        let viewportData = undefined;
+        if (isViewport(this.editorContext.modelRoot)) {
+            viewportData = <ViewportDataContainer>{
+                elementId: this.editorContext.modelRoot.id,
+                viewportData: {
+                    scroll: this.editorContext.modelRoot.scroll,
+                    zoom: this.editorContext.modelRoot.zoom
+                }
+            };
+        }
+        return viewportData;
+    }
+
+    protected async setViewportData(viewportData: ViewportDataContainer): Promise<void> {
+        if (this.actionDispatcher instanceof GLSPActionDispatcher) {
+            const restoreViewportAction = new SetViewportAction(viewportData.elementId, viewportData.viewportData, true);
```

Comment:
I was not 100% convinced about the animation as it feels a bit unclean for there to be something happening just after opening a file. However, without animation it sometimes looks like flickering which is worse in my opinion, so I am happy with the animation. Just wanted to note that down somewhere (I tried to make GIFs but they are not smooth enough to show it).